### DISCUSSION TITLE
cache is not invalidated after mutation for null field(in document cache)

### DISCUSSION THREAD

**Discussion by MH4GF** [urql-graphql/urql#2893] [Q&A]
Thanks for creating a great library. we're using urql in production.

## The problem that is happening

I want the query that returned null to be executed again after mutation, but it won't.

The following query, `data.order` may return null.

```typescript
const [{ data }] = useQuery({ query: `query { order { id name }`, context: { requestPolicy: 'network-only' } });

const [_, create] = useMutation(`mutation($name: String!) { createOrder(name: $name) }`);
const [_, update] = useMutation(`mutation($id: ID!, $name: String!) { updateOrder(id: $id, name: $name) }`);

const onClick = () => {
  if (data.order) {
    update({ id: data.order.id, name: 'newName'})
  } else {
    create({ name: 'newName' });
  }
};
```

If `data.order` has a value, the query is execute again after mutation, but not if `data.order` is null.
If `data.order` is null, stale in OperationResult remains false.

## What we tried

https://formidable.com/open-source/urql/docs/basics/document-caching/#adding-typenames
We thought this was relevant and set `additionalTypenames` to query, but the result did not change.

```typescript
const [{ data }] = useQuery({ query: `query { order { id name }`, context: { requestPolicy: 'network-only', additionalTypeNames: [`Order`]} });
```

Also, changing requestPolicy to another policy did not change the result.

## Additional Information

- next-urql: 3.3.3
- urql: 2.2.2

Please let me know if there is any other information you need.

thanks!

**Comment by JoviDeCroock** [comment]
If we can't derive the typename we can't invalidate the cache, mentioned in the gotcha's [here](https://formidable.com/open-source/urql/docs/basics/document-caching/#document-cache-gotchas) but for lists.

EDIT: I processed the second part of your message a bit late 😅 does `createOrder` have `Order` as a return type? Have you tried `additionalTypenames` without `network-only`? I assume so but just checking, normally adding the typename should make it invalidate will check how this can happen. Normally we just [append them](https://github.com/urql-graphql/urql/blob/main/packages/core/src/exchanges/cache.ts#L103-L105)

**Reply by MH4GF** [reply]
thanks for quick response!

> does createOrder have Order as a return type? 

yeah, here is my schema:

```graphql
type CreateOrderPayload {
  order: Order
  userErrors: [UserError!]! # defined in the my project
}

createOrder(
  input: OrderInput!
): CreateOrderPayload!
```

> Have you tried additionalTypenames without network-only?

Yes, I tried cache-first, etc., but the results did not change... 
The same was true when I set additionalTypeNames as the context for mutations.

I will try to see if I can prepare a minimal reproduction environment later.

**Reply by JoviDeCroock** [reply]
oh, the selection set of your mutation should be

```
const [_, create] = useMutation(`mutation($name: String!) { createOrder(name: $name) { order { id } } }`);
```

**Reply by MH4GF** [reply]
Sorry, I'm done the object retrieval with mutation, because if the data is non-null, it invalidates it 🤔

**Reply by JoviDeCroock** [reply]
No need to apologise 😅 did that fix it or? The reason behind this is that we need to find the Order typename so not selecting the object does not allow us to clear the cache

**Reply by MH4GF** [reply]
No, cache invalidate at creation has not yet been resolved...
For now, we are bypassing it by explicitly reexecuteQuery at creation.

**Reply by JoviDeCroock** [reply]
Have you tried the fix with the selection set though? Do you have the reproduction? We have a template in our issue template https://raw.githubusercontent.com/urql-graphql/urql/main/.github/ISSUE_TEMPLATE/bug_report.yaml

**Reply by MH4GF** [reply]
Thank you🙏🏻🙏🏻 , I will send a bug report later when the codesandbox is ready!

**Reply by MH4GF** [reply]
I'm working on preparing a minimum reproduction, but it has been properly invalidated by additionalTypeNames: [codesandbox](https://codesandbox.io/p/github/MH4GF/next-urql-document-cache-bug/main?selection=%5B%7B%22endColumn%22%3A38%2C%22endLineNumber%22%3A17%2C%22startColumn%22%3A38%2C%22startLineNumber%22%3A17%7D%5D&file=%2Fpages%2Findex.js&workspace=%257B%2522activeFileId%2522%253A%2522clbx558m400047nei1r5q04j5%2522%252C%2522openFiles%2522%253A%255B%2522%252FREADME.md%2522%255D%252C%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522gitSidebarPanel%2522%253A%2522COMMIT%2522%252C%2522spaces%2522%253A%257B%2522clbx567xn000x3b6kwk2xof93%2522%253A%257B%2522key%2522%253A%2522clbx567xn000x3b6kwk2xof93%2522%252C%2522name%2522%253A%2522Default%2522%252C%2522devtools%2522%253A%255B%257B%2522type%2522%253A%2522PREVIEW%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522key%2522%253A%2522clbx5m81700nn3b6jkr4iu7sm%2522%252C%2522isMinimized%2522%253Afalse%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522key%2522%253A%2522clbx5kn2z00573b6j4y88djsi%2522%252C%2522isMinimized%2522%253Afalse%257D%255D%257D%257D%252C%2522currentSpace%2522%253A%2522clbx567xn000x3b6kwk2xof93%2522%252C%2522spacesOrder%2522%253A%255B%2522clbx567xn000x3b6kwk2xof93%2522%255D%257D)
So there seems to be some kind of problem with our project. I will do some more research on this.
